### DISCUSSION TITLE
Is it possible to train a DRL-based robot controller with transformers using TRL?

### DISCUSSION THREAD

**Discussion by H-Hisamichi** [huggingface/trl#4079] [Q&A] [answered]
Hello everyone,

I am developing a DRL-based controller for a legged robot. However, for multi-purpose robots, I feel that using only an MLP has its limitations.
Therefore, I am considering applying transformers, but I am not sure whether TRL is suitable for integration with frameworks like Isaac Lab.
Since I am a beginner when it comes to transformers, I would greatly appreciate any advice or insights.

**Comment by demirciberk** [comment]
Short answer: TRL is not the right tool for this. TRL (Transformers Reinforcement Learning) is scoped to post-training text/language (and some vision-language) models with algorithms like PPOTrainer, GRPOTrainer, DPOTrainer, RLOOTrainer, etc. Its whole pipeline is built around tokenized text (prompt/completion), a PreTrainedModel from 🤗 Transformers, a PreTrainedTokenizerBase/ProcessorMixin, and reward computed from text generations (rule-based, a reward model, or preference data) — see the library overview: https://huggingface.co/docs/trl/index and the PPOTrainer API: https://huggingface.co/docs/trl/main/en/ppo_trainer

Confirmed facts (from official docs/source):
- TRL README states it trains "transformer language models" and lists its taxonomy as online methods (GRPO, RLOO, PPO, OnlineDPO, NashMD, XPO) and offline/preference methods (DPO, KTO, reward modeling) — all designed around token sequences and LM rollouts: https://github.com/huggingface/trl
- PPOTrainer (even in its current experimental form) takes model, ref_model, reward_model, value_model as PreTrainedModel/PreTrainedModelWrapper instances and a Dataset of text, not raw continuous observation vectors from a physics sim: https://huggingface.co/docs/trl/main/en/ppo_trainer
- TRL has no Env/Gym-style environment interface, no support for continuous action spaces, no vectorized-env rollout collection, and no integration hooks for Isaac Lab / Isaac Gym / MuJoCo / PyBullet in its docs or source.

Hypothesis / interpretation of your question:
- If what you actually want is "use a Transformer (attention-based) architecture as the policy/value network inside a legged-robot RL pipeline (e.g., PPO in Isaac Lab)," that's a valid and increasingly common idea (e.g., Decision Transformer, transformer-based locomotion policies), but it is an RL-for-control problem, not an LLM-post-training problem. You'd keep your existing DRL stack (Isaac Lab's own RL library, rl_games, rsl_rl, or skrl, which Isaac Lab already integrates: https://isaac-sim.github.io/IsaacLab/main/source/overview/reinforcement-learning/rl_frameworks.html) and just swap the MLP policy/value network for a Transformer/attention block (e.g., a small causal transformer over a history of proprioceptive observations) implemented in plain PyTorch, transformers.PretrainedConfig-free, or using a library like x-transformers. That gives you the "transformer for a robot controller" outcome without forcing your control loop through TRL's LM-training abstractions.

Minimal illustrative sketch (not tested, just to show the shape of the alternative approach — replace the policy MLP with a transformer encoder in your existing PPO implementation):
python
import torch, torch.nn as nn

class TransformerPolicy(nn.Module):
    def init(self, obs_dim, act_dim, d_model=128, n_layers=2, n_heads=4, seq_len=16):
        super().init()
        self.embed = nn.Linear(obs_dim, d_model)
        self.pos = nn.Parameter(torch.zeros(1, seq_len, d_model))
        layer = nn.TransformerEncoderLayer(d_model, n_heads, batch_first=True)
        self.encoder = nn.TransformerEncoder(layer, n_layers)
        self.act_head = nn.Linear(d_model, act_dim)

    def forward(self, obs_seq):  # obs_seq: (batch, seq_len, obs_dim)
        x = self.embed(obs_seq) + self.pos[:, :obs_seq.size(1)]
        x = self.encoder(x)
        return self.act_head(x[:, -1])  # action from latest timestep

This would plug into the standard PPO update loop used by rsl_rl/rl_games/skrl, not into TRL's PPOTrainer.

Suggested next step: clarify whether the goal is (a) fine-tuning an actual language/vision-language model to act as a high-level planner/controller that TRL could post-train, in which case TRL is applicable but you'd need a text or multimodal interface to the robot; or (b) using transformer architectures purely as the function approximator inside a classic DRL loop, in which case look at Isaac Lab's RL framework docs and papers like "Decision Transformer" (https://arxiv.org/abs/2106.01345) instead of TRL.

**Reply by H-Hisamichi** [reply]
Hello @demirciberk 

Thank you so much for sharing the current state of TRL — that's really helpful.
Things have moved on a bit since I opened this thread: I'm now working on combining multiple expert policies, either through policy distillation or action mixing. I'll look through the resources you linked right away.

Thanks again for the advice!

**Reply by demirciberk** [reply, marked as answer]
Glad it helped! If you're set, marking this comment as the answer helps others find it faster.
